File: CPP/6_zigZagConversion/6_zigZagConversion.cpp
// 6_zigZagConversion.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include <iostream>
#include <string>
#include <vector>
#include <map>
using namespace std;

string zigZagConversion(string src, int num)
{
	int row = 0, flag = 1;
	map<int, vector<char>> rowMap;
	for (int i = 0; i < (int)src.length(); i++)
	{
		rowMap[row].push_back(src[i]);
		if (flag == 1)
		{
			if (++row == num - 1)
			{
				flag = 0;
			}
		}
		else
		{
			if (--row == 0)
			{
				flag = 1;
			}
		}
	}

	string dst;
	map<int, vector<char>>::iterator iter1 = rowMap.begin();
	for (; iter1 != rowMap.end(); iter1++)
	{
		vector<char>::iterator iter2 = iter1->second.begin();
		for (; iter2 != iter1->second.end(); iter2++)
		{
			dst += *iter2;
		}
	}
	return dst;
}

int _tmain(int argc, _TCHAR* argv[])
{
	string str = "PAYPALISHIRING";
	cout << zigZagConversion(str, 4) << endl;
	system("pause");
	return 0;
}

File: Go/6_zigZagConversion/6_zigZagConversion.go
package q006

import (
	"fmt"
)

func zigZagConversion(src string, row int) string {
	if row == 1 {
		return src
	}

	if row > len(src) {
		row = len(src)
	}

	strs := make([]string, row)
	curr := 0
	godown := true
	for i := 0; i < len(src); i++ {
		strs[curr] += string(src[i])

		if godown {
			curr++
		} else {
			curr--
		}

		if curr == 0 || curr == row - 1 {
			godown = !godown
		}
	}

	dst := ""
	for i := 0; i < len(strs); i++ {
		dst += strs[i]
	}
	return dst
}

// Test ...
func Test() {
	fmt.Println("6_zigZagConversion begin")
	str := "PAYPALISHIRING"
	fmt.Println(zigZagConversion(str, 4))
	fmt.Println("6_zigZagConversion end")
}
